### PR TITLE
Unify most of the test apps into `on device unit test app`

### PR DESCRIPTION
**Pull request by opacam**
So depending on the requested recipes at built time, a different test app will be ran (as well as different unit tests). We also set this *unified test app* the default for our `CI` tests (since we remove the `testapp_sqlite3_openssl` ... alongside with some more test apps).

There are 3 app modes:
  - **kivy** (if kivy detected in `app_requirements.txt`)
  - **flask** (if flask detected in `app_requirements.txt`)
  - **no-gui** (if above cases are not taken)

In all the cases mentioned the unittests will be run. In the two first use cases (kivy and flask) the result of the unit tests can be seen via the app. If an `no-gui` app is built, then the results must be checked via `adb logcat`.

Also be aware that this PR adds **more artifacts for gh-actions** via an `matrix`, so now we will have `on device unit test app` for:
   - **Python3 & arm64-v8a** *(tested on device)*
   - **Python3 & armeabi-v7a** *(tested on device)*
   - **Python3 & x86_64**
   - **Python3 & x86**
   - ~~**Python2 & arm64-v8a** *(tested on device)*~~
   - ~~**Python2 & armeabi-v7a** *(tested on device)*~~
   - ~~**Python2 & x86_64**~~

**A quick view of the situation:**
   - the removed `testapp_keyboard` and  `testapp_service`, now lives in the `on device unit test app`
   - the testapps for `pillow`, `matplotlib` and `encryption` has been moved into the unit tests (so we can also test those recipes via the flask app)
  - the testapp for `flask` has been also moved/reworked into the new `on device unit test app`, this allow us to share the resources with the kivy app (which are the same).
  - Furthermore,  we generate a file at build time `app_requirements.txt` which will hold the supplied recipes at build time. This file will allow us to test these recipes via our unittests. Just noting that the recipes we want to test, must be explicitly mentioned. Also should be mentioned that this feature its only available for **direct build with p4a** (when `setup_test_app.py` is ran)...so **buildozer builds will build an basic test app**, no matter what recipes are introduced into `buildozer.spec` file, because the  `app_requirements.txt` will only be created if the `setup_test_app.py` file is ran.

**Notes:**
  - @inclement, I included you as author of this thing (alongside myself), I think that we are the authors of the removed test apps and you are also the author of the base code for the `on device unit test app` (@AndreMiras and *anyone looking into this*, I think that you are not involved with these test apps, please, let me know if I'm mistaken and I will include you as author as well).
  - I avoided python3's specific code in purpose, so the new test app could be also run from python2 (this way we can remove the python2 test apps). BTW, we only have one `import` to add python2 compatibility to this new unit test app...imho...I thinks it's worth it :wink:
  - I labeled this `WIP` because:
    + I want to add a couple of commits, in order to force to generate the flask testapp via gh-actions and restore it to kivy, so this way, you guys can see the result of this `flask unittest testapp` without building anything, for now you can download the `kivy testapp` at gh-actions
    + I have mixed feelings about this PR...maybe the whole thing it's a crazy idea of mine...mmmm...let me know if this is the case and I will close this, no problem at all
  - I added a font file (it's very light...less than 10Kb and it's  [free for commercial/personal use](https://www.wfonts.com/font/blanka)). This font has two purposes:
    + used for some titles in kivy/flask apps
    + to perform the unittests for the Pillow recipe
  - @inclement, the testapp for keyboard that I deleted and included at the new unit test app...I'm not sure if its working as expected...it would be great that you could download one of the generated tests apps via gh-actions and give your thoughts, @AndreMiras, it would be awesome to have yours as well.
  - I did not test the `buildozer` build, anyone could test it?
  - Closes #2073

**Some future improvements we could make:**
  - remove/integrate some more test apps. Merging this PR, we will only have the following test apps:
      + `testapp_vispy`
      + `testlauncher_setup`
      + `testlauncherreboot_setup`
  - make gh-actions create a new artifact for `rebuild updated recipes` (I have some job done in this point but not included at here)
  - add tests for all the recipes we have (we have a basic set but we could go further)
  - Improve design and features, specially for the flask app (yea..I put more love with the kivy app :smile:)
  - test the artifacts we generate with gh-actions on device...@AndreMiras, you were working with that for travis...maybe now it will be easier to do that with gh-actions?
  - you tell me...

### CODE REVIEW

File: testapps/on_device_unit_tests/test_app/main.py [hunk 1]
```diff
@@ -1,45 +1,97 @@
...
+have an ui, or not, depending on requirements as well.
 
-class PythonTestMixIn(object):
+For now, we contemplate three possibilities:
+  -  A kivy unittest app (sdl2 bootstrap)
+  -  A unittest app (webview bootstrap)
+  -  A non-gui unittests app
 
-    module_import = None
+If you install/build this app via the `setup_test_app.py` file, an file named
```
Comment (AndreMiras) [inline testapps/on_device_unit_tests/test_app/main.py]: s/an file/a file/

File: testapps/on_device_unit_tests/test_app/main.py [hunk 2]
```diff
@@ -1,45 +1,97 @@
...
 
-    def test_import_module(self):
-        """Test importing the specified Python module name. This import test
-        is common to all Python modules, it does not test any further
-        functionality.
-        """
-        self.assertIsNotNone(
-            self.module_import,
-            'module_import is not set (was default None)')
+.. note:: This app it's made to be working on desktop and on an android device.
```
Comment (AndreMiras) [inline testapps/on_device_unit_tests/test_app/main.py]: s/it's made/is made/

File: testapps/on_device_unit_tests/test_app/main.py [hunk 3]
```diff
@@ -1,45 +1,97 @@
...
+          Be aware that some of the functionality of this app will only work on
+          an android device. Also should be mentioned that it will run for
+          python2 and python3.
```
Comment (AndreMiras) [inline testapps/on_device_unit_tests/test_app/main.py]: I know we still support target Python2, but I would keep it a secret by removing that last sentence :stuck_out_tongue:

File: testapps/on_device_unit_tests/test_app/main.py [hunk 4]
```diff
@@ -1,45 +1,97 @@
...
 
-        importlib.import_module(self.module_import)
+.. tip:: you can write more unit tests at `tests/test_requirements.py` and test
+         these on desktop just by editing the file `app_requirements.txt`,
+         which should be located at the same location than this file. This
+         `app_requirements.txt` file, it's autogenerated when the
+         `setup_test_app.py` is ran, so  in certain circumstances, you may need
```
Comment (AndreMiras) [inline testapps/on_device_unit_tests/test_app/main.py]: s/so  in/so in/

**Comment by inclement** [comment]
Thanks @opacam. Please let me review before merging :)

**Comment by opacam** [comment]
@inclement, I fixed the @AndreMiras comments and also made some additional changes (hope that the commits are clear enough). I leave this **WIP** awaiting your review (@AndreMiras...and yours again if you want) you probably want to change something, so let me know or make the changes that you want directly in here (you should be able to do so, because I always mark the checkbox *allow edit from maintainers*), whatever it's fine to me.

Also, here you have the download links for the generated apks via gh-actions (you can collect yourself this via the artifacts dropdown of the gh-actions corresponding job, but I collected the relevant ones for you guys :wink:):
  - [Flask apk (python3 - arm64-v8a)](https://github.com/kivy/python-for-android/suites/377778226/artifacts/875781) (collected from commit 0ed4d70)
  - [Flask apk (python3 - armeabi-v7a)](https://github.com/kivy/python-for-android/suites/377778226/artifacts/875782) (collected from commit 0ed4d70)
  - [kivy apk (python3 - arm64-v8a)](https://github.com/kivy/python-for-android/suites/386781165/artifacts/973343) (collected from the last gh-actions job, 896c819) 
  - [kivy apk (python3 - armeabi-v7a)](https://github.com/kivy/python-for-android/suites/386781165/artifacts/973344) (collected from the last gh-actions job, 896c819)

**Note:** Since I consider this very useful for us (specially in combination with gh-actions) and I don't want to rush you with the review, I'm planning to open a draft pull request with this branch `squashed and merged` on top of develop branch , so I can play a little with the requirements of the `on device unit test app`, just to test some recipes which I know that are not working (and I will try to fix it  there), so we all can see this thing in action and maybe extract some fixes for the non working recipes.

A quick "run in desktop" tutorial
------------------------------------------
If you just want to see how it will look the kivy/flask apps from your PC, you can run the main.py file with the proper python environment/installation. If you don't make any build nor touch `setup_test_app.py`, then the kivy app should run (of course you need kivy installed), if you edit the mentioned file replacing kivy for flask in requirements (setup_test_app.py or via cli) and make a build as we used to do with the removed `setup_testapp_python3_sqlite_openssl.py`, well...flask app should raise (you need flash installed and a browser to see the results)

**Hack**: you can avoid do the build if you create the file `app_requirements.txt` file inside `testapps/on_device_unit_tests/test_app` folder with the requirements you want to test. The content of such file will look like as following (without touching `setup_test_app.py` and performing a build of an apk):
```
sqlite3
libffi
openssl
pyjnius
kivy
python3
requests
```

File: testapps/on_device_unit_tests/setup_test_app.py [hunk 5]
```diff
@@ -0,0 +1,92 @@
+"""
+This is the `setup.py` file for the `on device unit test app`.
+
+In this module we can control how will be built our test app. Depending on
+our requirements we can build an kivy, flask or a non-gui app. We default to an
+kivy app, since the python-for-android project its a sister project of kivy.
+
+The parameter `requirements` it's crucial to determine the unit tests we will
```
Comment (AndreMiras) [inline testapps/on_device_unit_tests/setup_test_app.py]: `it's crucial` -> `is crucial`

**Comment by opacam** [comment]
Ok, I fixed the merge conflicts and restored the `setup_testapp_python3_openssl_sqlite.py`. BTW...I also use it, but lately I began to use the  `on_device_unit_test_app` as a replacement because I find it more versatile since it does the same than `setup_testapp_python3_openssl_sqlite.py` but a little more :smile:. Anyway, I have no problem on leaving it in the project, after all...it has been a lot of helpful to me :wink:

**Pull request merged**